linera-core/src/client.rs

```diff
@@ -2203,42 +2203,91 @@ where
+
+        async fn await_while_polling<F: FusedFuture>(
+            future: F,
+            background_work: impl FusedFuture<Output = ()>,
+        ) -> F::Output {
+            tokio::pin!(future);
+            tokio::pin!(background_work);
+            futures::select! {
+                _ = background_work => future.await,
+                result = future => result,
+            }
+        }
+
         let mut senders = HashMap::new(); // Senders to cancel notification streams.
         let mut guard = self.lock().await;
         let notifications = guard.subscribe().await?;
-        let notifications2 = guard.subscribe().await?;
+        let (abortable_notifications, abort) = stream::abortable(guard.subscribe().await?);
         if let Err(error) = guard.synchronize_from_validators().await {
             error!("Failed to synchronize from validators: {}", error);
         }
         drop(guard);
-        let (mut notifications, abort) = stream::abortable(notifications);
-        if let Err(error) = self.update_streams(&mut senders).await {
-            error!("Failed to update committee: {}", error);
-        }
+
+        // Beware: if this future ceases to make progress, notification processing will
+        // deadlock, because of the issue described in
+        // https://github.com/linera-io/linera-protocol/pull/1173.
+
+        // TODO(#2013): replace this lock with an asychronous communication channel
+
+        let mut process_notifications =
+            future::OptionFuture::from(match self.update_streams(&mut senders).await {
+                Ok(handler) => Some(handler),
+                Err(error) => {
+                    error!("Failed to update committee: {error}");
+                    None
+                }
+            })
+            .map(drop)
+            .fuse();
+
         let this = self.clone();
-        tokio::spawn(async move {
-            while let Some(notification) = notifications.next().await {
-                if matches!(notification.reason, Reason::NewBlock { .. }) {
-                    if let Err(error) = this.update_streams(&mut senders).await {
-                        error!("Failed to update committee: {}", error);
+        let update_streams = async move {
+            let mut abortable_notifications = abortable_notifications.fuse();
+
+            loop {
+                let Some(notification) =
+                    await_while_polling(abortable_notifications.next(), &mut process_notifications)
+                        .await
+                else {
+                    break;
+                };
+
+                if let Reason::NewBlock { .. } = notification.reason {
+                    let update_result = await_while_polling(
+                        this.update_streams(&mut senders).fuse(),
+                        &mut process_notifications,
+                    )
+                    .await;
+                    match update_result {
+                        Ok(handler) => handler.await,
```

Comment:
Isn't the `handler` a task that keeps processing notifications from added validators forever? Then we can't `await` it, but would need to somehow add it to the `process_notifications` background task.

Reply:
Yep, I think that's the current bug :)